File: examples/hello_pubsub/hello_pubsub_py/hello_pubsub_py/hello_publisher.py
#!/usr/bin/env python3

'''
    A minimal ROS2 python node that publishes a string every second.

    Basic structure for ROS2 node publisher:
    - Initialise rclpy
    - Create a node
    - Create the publisher
    - Create a timer to publish at specific intervals
    - Keep it alive with spin
    - Shutdown

    BEST_EFFORT and VOLATILE have been selected for the QOS_PROFILE, as the
    node keeps publishing at fixed intervals and losing messages it's fine.

    * If you want reliability change BEST_EFFORT to RELIABLE.
    * If you want persisting samples for late-joining subscriptions,
      change VOLATILE to TRANSIENT_LOCAL.

    Author: Ioannis Selinis <selinis.g@gmail.com> 2025
'''


import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy, DurabilityPolicy
from example_interfaces.msg import String


# https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Quality-of-Service-Settings.html
QOS_PROFILE = QoSProfile(
    reliability=ReliabilityPolicy.BEST_EFFORT,      # RELIABLE or BEST_EFFORT
    durability=DurabilityPolicy.VOLATILE,           # VOLATILE or TRANSIENT_LOCAL
    history=HistoryPolicy.KEEP_LAST,                # KEEP_ALL or KEEP_LAST
    depth=1                                         # Used only with KEEP_LAST
)


class HelloPublisher(Node):
    """The HelloPublisher class that inherits from rclpy.node.Node.
    Initialises with a counter and a publisher and creates a timer
    that triggers once per second to publish a message.
    """
    def __init__(self):
        super().__init__(node_name='hello_publisher')

        self.counter = 0
        # ~/hello_world would publish it as /hello_publisher/hello_world
        self.publisher = self.create_publisher(msg_type=String,
                                               topic='hello_world',
                                               qos_profile=QOS_PROFILE)
        self.timer = self.create_timer(timer_period_sec=1.0,
                                       callback=self.publisher_callback)


    def publisher_callback(self):
        """Publish the message once per second."""
        msg = String()
        msg.data = f'Hello counter: {self.counter}'
        self.publisher.publish(msg=msg)
        self.counter += 1


def main(args=None) -> None:
    """Entry Point for our node"""
    rclpy.init(args=args)   # Initialise rclpy (ROS communications)
    node = HelloPublisher()
    rclpy.spin(node=node)   # Keep node alive
    node.destroy_node()     # Destroy node
    rclpy.shutdown()        # Shutdown the executor


if __name__ == '__main__':
    main()

File: examples/hello_pubsub/hello_pubsub_cpp/src/hello_publisher.cpp
/**
 *  A minimal ROS2 cpp node that publishes a message every second.
 *
 *  Basic structure for ROS2 node:
 *  - Initialise rclpy
 *  - Create a node
 *  - Create the publisher
 *  - Create a timer to publish at specific intervals
 *  - Keep it alive with spin
 *  - Shutdown
 *
 *  BEST_EFFORT and VOLATILE have been selected for the QOS_PROFILE, as the
 *  node keeps publishing at fixed intervals and losing messages it's fine.
 *
 *  * If you want reliability change BEST_EFFORT to RELIABLE.
 *  * If you want persisting samples for late-joining subscriptions,
 *    change VOLATILE to TRANSIENT_LOCAL.
 *
 *  Author: Ioannis Selinis <selinis.g@gmail.com> 2025
 */

#include <rclcpp/rclcpp.hpp>
#include "example_interfaces/msg/string.hpp"


static const rmw_qos_profile_t rmw_qos_profile_latch =
{
    RMW_QOS_POLICY_HISTORY_KEEP_LAST,
    1,
    RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT,
    RMW_QOS_POLICY_DURABILITY_VOLATILE,
    RMW_QOS_DEADLINE_DEFAULT,
    RMW_QOS_LIFESPAN_DEFAULT,
    RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT,
    RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT,
    false
};


class HelloPublisher : public rclcpp::Node
{
    public:
        HelloPublisher() : Node("hello_c_publisher")
        {
            auto qos = rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile_latch), rmw_qos_profile_latch);
            m_timer = this->create_wall_timer(std::chrono::seconds(1),
                                              std::bind(&HelloPublisher::timerCallback, this));
            m_publisher = this->create_publisher<example_interfaces::msg::String>("hello_world", qos);
            RCLCPP_INFO(this->get_logger(), "Node has started");
        }

    private:
        void timerCallback()
        {
            auto msg = example_interfaces::msg::String();
            msg.data = "Hello counter: " + std::to_string(m_counter);
            m_publisher->publish(msg);
            m_counter++;
        }
        rclcpp::TimerBase::SharedPtr m_timer;
        rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr m_publisher;
        int m_counter = 0;
};


int main (int argc, char **argv)
{
    rclcpp::init(argc, argv);                       // Initialise rclcpp (ROS communications)
    auto node = std::make_shared<HelloPublisher>(); // Smart/Shared pointer
    rclcpp::spin(node);                             // Keep node alive
    rclcpp::shutdown();                             // Shutdown executor
}
